ocamlc -config: new -config-var option to print specific configuration variables

gasche opened the pull request:
The proposed behavior of `-config-var s` is as follows:
- if `s` is an existing configuration variable, print its value as
  a string, a newline, and exit with a success return value (0)
- if `s` is not an existing configuration variable, print nothing
  and exit with a failure return value (non-0)

This gives two ways to distinguish success from failure to script
authors: the return code, and counting the lines of output.

The following alternative behaviors were considered:

- We could *not* print a newline after the configuration value, which
  might arguably be useful in some configuration scenarios. The
  utility is doubtful as Unix shell expansions
  `$(ocamlc -config-var ...)` remove trailing newlines, and it prevents
  the line-count-based failure detection (an empty configuration variable
  would also print nothing).

- We could print a message on the error output if the configuration
  variable does not exist. This is clearer to a human user, but it is
  annoying for scripts if they forget to silence the error output and
  get their output mixed with our error messages. The main use of this
  new feature is for scripting purposes.

gasche committed:
ocamlc -config: new -config-var option to print specific configuration variables

The proposed behavior of `-config-var s` is as follows:
- if `s` is an existing configuration variable, print its value as
  a string and exit with a success return value (0)
- if `s` is not an existing configuration variable, print nothing
  and exit with a failure return value (non-0)

Note that we do not print a newline after the value of the
configuration variable. In particular, if the value is an empty
string, the output is undistinguishable from the output for
non-existing variables, the return value has to be considered instead.

The following alternative behaviors were considered:

- We could print a newline after the configuration value, which
  would let users distinguish empty values from non-existing variables
  by counting the lines of output, and would also be more pleasant for
  users invoking the option from the command-line. However, the way
  bash works on Windows means that $(ocamlc -config-var foo) would keep
  a trailing \r in its output, and portable scripts would have to use
  $(ocamlc -config-var foo | tr -d '\r') instead, which is a pain.
  (This issue was pointed out by David Allsopp)

- We could print a message on the error output if the configuration
  variable does not exist. This is clearer to a human user, but it is
  annoying for scripts if they forget to silence the error output and
  get their output mixed with our error messages. The main use of this
  new feature is for scripting purposes.

gasche commented:
(cc @dra27, @dbuenzli)

shindere commented:
Gabriel Scherer (2017/10/14 17:54 +0000):
> - We could print a message on the error output if the configuration
>   variable does not exist. This is clearer to a human user, but it is
>   annoying for scripts if they forget to silence the error output and
>   get their output mixed with our error messages. The main use of this
>   new feature is for scripting purposes.

There is the question of how to deprecate variables that we may want to
address here. It seems to me it could be done precisely by printing a
deprecation message to stderr.

dra27 commented on utils/config.mlp:
```diff
@@ -148,10 +148,16 @@ let systhread_supported = %%SYSTHREAD_SUPPORT%%;;
+  let p name valu = (name, String valu) in
+  let p_int name valu = (name, Int valu) in
+  let p_bool name valu = (name, Bool valu) in
+[
```
As this being updated, could we do s/valu/value ?

dra27 commented on utils/config.mlp:
```diff
@@ -204,6 +209,30 @@ let print_config oc =
 
+let print_config oc =
+  let print (name, valu) =
+    Printf.fprintf oc "%s: %a\n" name print_config_value valu in
```
Similarly here s/valu/value

dra27 commented on utils/config.mlp:
```diff
@@ -204,6 +209,30 @@ let print_config oc =
+  match List.assoc_opt name configuration_variables with
+  | None -> None
+  | Some v ->
+      let s = match v with
```
It's a ~small~microscopic nit, but `valu`/`value` is used elsewhere to refer to a value, but `v` here

dra27 commented on utils/misc.ml:
```diff
@@ -730,3 +730,16 @@ module MakeHooks(M: sig
+
+let show_config_variable_and_exit var =
+  match Config.config_var var with
+  | Some valu ->
```
Same here with s/valu/value!

dra27 commented on utils/misc.ml:
```diff
@@ -730,3 +730,16 @@ module MakeHooks(M: sig
+let show_config_variable_and_exit var =
+  match Config.config_var var with
+  | Some valu ->
+      print_endline valu;
```
Why do this - **not** doing this will be amazingly wonderful for compatibility between Windows/Cygwin and the rest of the world where shell scripting is concerned (no more `tr -d '\r'` all over the place)

gasche replied:
You mean why include a line break at the end instead of using `print_string`?

dra27 replied:
Yes, indeed - sorry for being somewhat less than clear!

dra27 reviewed (commented):
Thanks for doing this! I agree with @shindere that it would be nice to work in a mechanism for deprecation, which would allow `bytecomp_c_compiler` and `native_c_compiler` to display a deprecation warning on stderr.

If that is done too, this is a pretty low-risk change - might inclusion in 4.06.0 be safe to do?

gasche commented:
I don't like `valu` but I'm not fond of `value` either, so I may just use `v` if that tickles.

shindere commented:
Gabriel Scherer (2017/10/20 05:11 -0700):
> I don't like `valu` but I'm not fond of `value` either, so I may just
> use `v` if that tickles.

@gasche why don't you like value, actually?

gasche commented:
I've written enough code in revised syntax to see it as a keyword.

shindere commented:
Gabriel Scherer (2017/10/20 13:14 +0000):
> I've written enough code in revised syntax to see it as a keyword.
Which it isn't in vanilla ocaml, right?

"vaue" really seems like the right name to use here, IMHO.

gasche commented:
`v` will be just fine.

shindere commented:
Gabriel Scherer (2017/10/20 06:35 -0700):
> `v` will be just fine.

In a context where there are both variables and values, I don't agree
but won't insist further. :)

dbuenzli commented:
> In a context where there are both variables and values

Unless the terminology is already pervasive in the code you could use the configuration key terminology rather than configuration variable, and thus `k` and `v`.

gasche commented:
I rebased the PR with the variable renaming (and against trunk). I went for `x` and `v` instead of `k` and `v` because then I would have had to rename `-config-var` into `-config-key` and I found I like "configuration variable and value" better than "configuration key and value" (but let your preference be known if it differs).

I am procrastinating against removing the newline for no very good reason. Deep down I like printing a newline (for testing the feature from the command-line) and I find it *extremely* frustrating that bash on Windows does not strip the trailing `\r\n` away and that I have to change my code for this. I also spent some time staring at David's `tr -d` calls all over the build system in pain and disbelief. I plan to wait until the burning fades away and do the reasonable thing of using `print_string`.

xavierleroy commented:
> I am procrastinating against removing the newline for no very good reason

I think you should bite the bullet and do it.  We'll see how users react to this.

> I like printing a newline (for testing the feature from the command-line)

If we want to get fancy (but not in 4.06) we could print a newline if the output is a terminal...  Now that we have a "isatty" function in the runtime, and that it is almost working, we could use it more widely.

gasche commented:
I rebased the PR to remove the newline, but I remain unsatisfied. Could someone that knows more about shell scripting than I do tell me how they would typically (and correctly) use this new `-config-var` option in a shell script? I must admit that I don't understand how to test for the return code of the command in a nice/concise way.

(Edit: a previous version of this message had an incorrect shell script that didn't work below. Fixed.)

The best I found is

```shell
VALUE=$(ocamlc -config-var $VARIABLE)
if [ $? -ne "0" ]
then echo "variable does not exist"
else echo "variable does exist and its value is \"$VALUE\""
fi
```

I'm thinking that maybe I should also add a `-config-has-var <variable>` command that would output `true` or `false`, because shell scripting is just too hard.

gasche commented:
This appears to work and it does look reasonably nice (in particular, contrarily to the `$?` business, it is not too fragile if several calls to `ocamlc -config-var` are made in sequence).

```shell
if VALUE=$(ocamlc -config-var $VARIABLE)
then echo "variable does exist and its value is \"$VALUE\""
else echo "variable does not exist"
fi
```

dra27 commented:
I think you may be virtuously applying good software engineering principles to the evil world of shell scripting 🙂. I would anticipate doing something like:

```bash
PREFIX=$(ocamlc -config-var prefix 2>&1 || ocamlc -config | tr -d '\015' | sed -ne "s/prefix: //p")
```

and expect things to fail noisily in the strange event that `PREFIX` is `""`. If I were writing something which requires an OCaml which has this feature, then I wouldn't expect to test the exit code at all, I'd just expect `ocamlc -config-var prefix` to have worked (having previously checked `ocamlc -vnum`, etc.)

gasche commented:
I rebased, I believe all comments have been taken into account, and I added manpage and manual documentation.

damiendoligez commented on otherlibs/dynlink/Makefile:
```diff
@@ -38,7 +38,7 @@ endif
 
 COMPILEROBJS=\
-  ../../utils/misc.cmo ../../utils/config.cmo \
+  ../../utils/config.cmo ../../utils/misc.cmo \
```
This makes me slightly nervous. Maybe `show_config_and_exit` and `show_config_var_and_exit` should be in `config.ml` rather than `misc.ml`?

gasche replied:
I think that `config` should remain a "pure data" module as much as possible.

I don't think we should add logic to it. Maybe I could create a separate `config_misc` module depending on `config`?

On the other hand, I think that `config` should really be at the bottom of the module dependency graph (especially if we commit to keeping it a pure data module), so I think the change is right in the long term -- although I agree it makes us a bit nervous.

damiendoligez replied:
I agree that `Config` should be a pure data module and that pure data modules should be at the bottom of the graph, so this is OK after all.

damiendoligez commented on driver/main_args.ml:
```diff
@@ -64,6 +64,12 @@ let mk_config f =
+let mk_config_var f =
+  "-config-var", Arg.String f,
+  " Print the value of a configuration variable, a newline, and exit \
+    (print nothing and exit with error value if the variable does not exist)"
```
Could you please cut this message with `\n\` and not make the compiler's help output look like shit (any more than it already does)?

gasche replied:
@damiendoligez could I implement an `Arg.wrap : ~line_length:int -> specs -> specs` function instead of manually doing the job of choosing the right cut place depending on the option name and alignment choice?

damiendoligez replied:
Sure, go ahead.

gasche replied:
In the end I did the simple thing of adding turning the end-of-line `\` into `\n     \`.

damiendoligez reviewed (changes requested):
Looks pretty good but I have two remarks.

gasche commented:
I had forgotten that this hadn't been merged yet (I thought of it because of #1691). I will do the short fix of adding `\n`s in the help output, and propose this for merging again.

gasche commented:
@damiendoligez, I believe that your changes have been addressed. Would you approve the PR? My plan would be to merge into trunk (not 4.07) once the CI passes, to be done with this change.

damiendoligez commented on manual/manual/cmds/unified-options.etex:
```diff
@@ -152,9 +152,19 @@ results in slightly smaller but slightly slower programs. The default is to
+\item["-config-var" \var{var}]
+Print the value of a specific configuration variable from the
+"-config" output, then exit. If the variable does not exist, the exit
+code is non-zero. This option is only available since OCaml 4.07,
```
You'll have to change this to 4.08 :-(

gasche replied:
Thanks for the catch, I changed it and rebased the PR.
